lib/src/view/analysis/tree_view.dart

```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
           ),
         SliverFillRemaining(
           hasScrollBody: false,
-          child: Padding(
-            padding: const EdgeInsets.symmetric(vertical: 10, horizontal: 10),
-            child: Wrap(
-              spacing: kInlineMoveSpacing,
-              children: moveWidgets,
+          child: _PgnTreeView(
+            root: root,
+            rootComments: rootComments,
+            params: (
+              shouldShowAnnotations: shouldShowAnnotations,
+              shouldShowComments: shouldShowComments,
+              currentMoveKey: currentMoveKey,
+              currentPath: currentPath,
+              notifier: ref.read(ctrlProvider.notifier),
             ),
           ),
         ),
       ],
     );
   }
+}
+
+/// A group of parameters that are passed through various parts of the tree view
+/// and ultimately evaluated in the InlineMove widget. Grouped in this record to improve readability.
+typedef _PgnTreeViewParams = ({
+  UciPath currentPath,
+  bool shouldShowAnnotations,
+  bool shouldShowComments,
+  GlobalKey currentMoveKey,
+  AnalysisController notifier,
+});
+
+/// True if the side line has no branching and is less than 6 moves deep.
+bool _displaySideLineAsInline(ViewBranch node, [int depth = 0]) {
+  if (depth == 6) return false;
+  if (node.children.isEmpty) return true;
+  if (node.children.length > 1) return false;
+  return _displaySideLineAsInline(node.children.first, depth + 1);
+}
+
+bool _hasNonInlineSideLine(ViewNode node) =>
+    node.children.length > 2 ||
+    (node.children.length == 2 && !_displaySideLineAsInline(node.children[1]));
+
+Iterable<List<ViewNode>> _mainlineParts(ViewRoot root) =>
+    [root, ...root.mainline]
+        .splitAfter(_hasNonInlineSideLine)
+        .takeWhile((nodes) => nodes.firstOrNull?.children.isNotEmpty == true);
+
+class _PgnTreeView extends StatefulWidget {
+  const _PgnTreeView({
+    required this.root,
+    required this.rootComments,
+    required this.params,
+  });
+
+  final ViewRoot root;
+
+  final IList<PgnComment>? rootComments;
+
+  final _PgnTreeViewParams params;
+
+  @override
+  State<_PgnTreeView> createState() => _PgnTreeViewState();
+}
+
+typedef _TreeViewPart = ({
+  List<Widget> lines,
+  bool containsCurrentMove,
+});
+
+class _PgnTreeViewState extends State<_PgnTreeView> {
+  List<List<ViewNode>> mainlineParts = [];
```

Comment:
As I'm reading this for the first time, I'm not sure to understand exactly what `mainlineParts` and `treeParts` are for. The types are not enough to self explain.

Some comments explaining the logic would be welcome where needed imo.